sessions/test_js_challenge.py

```diff
@@ -2,16 +2,133 @@
+        self.assertEqual(curr_responses, len(client.responses))
+        self.assertTrue(client.connection_is_closed())
+
+    @abc.abstractmethod
+    def prepare_js_templates(self):
+        pass
+
+    def start_all(self):
+        self.prepare_js_templates()
+        self.start_all_servers()
+        self.start_tempesta()
+        self.start_all_clients()
+        self.deproxy_manager.start()
+        self.assertTrue(self.wait_all_connections(1))
+
+    def expect_restart(self, client, req, status_code, last_cookie):
+        """We tried to pass JS challenge, but the cookie we have was generated
+        too long time ago. We can't pass JS challenge now, but Tempesta
+        doesn't block us, but gives second chance to pass the challenge.
+
+        Expect a new redirect response with new sticky cookie value.
+        """
+        resp = self.client_send_req(client, req)
+        self.assertEqual(resp.status, '%d' % status_code,
+                         "unexpected response status code")
+        c_header = resp.headers.get('Set-Cookie', None)
+        self.assertIsNotNone(c_header,
+                             "Set-Cookie header is missing in the response")
+        match = re.search(r'([^;\s]+)=([^;\s]+)', c_header)
+        self.assertIsNotNone(match,
+                             "Can't extract value from Set-Cookie header")
+        new_cookie = (match.group(1), match.group(2))
+        self.assertNotEqual(last_cookie, new_cookie,
+                            "Challenge is not restarted")
+
+    def process_js_challenge(self, client, host, delay_min, delay_range,
+                             status_code, expect_pass, req_delay,
+                             restart_on_fail=False, check_redir=False):
+        """Our tests can't pass the JS challenge with propper configuration,
+        enlarge delay limit to not recommended values to make it possible to
+        hardcode the JS challenge.
+        """
+        req = ("GET / HTTP/1.1\r\n"
+               "Host: %s\r\n"
+               "Accept: text/html\r\n"
+               "\r\n" % (host))
+        resp = self.client_send_req(client, req)
+        self.assertEqual(resp.status, '%d' % status_code,
+                         "unexpected response status code")
+        c_header = resp.headers.get('Set-Cookie', None)
+        self.assertIsNotNone(c_header,
+                             "Set-Cookie header is missing in the response")
+        match = re.search(r'([^;\s]+)=([^;\s]+)', c_header)
+        self.assertIsNotNone(match,
+                             "Cant extract value from Set-Cookie header")
+        cookie = (match.group(1), match.group(2))
+
+        uri = resp.headers.get('Location', None)
+        self.assertIsNotNone(uri,
+                             "Location header is missing in the response")
+        if check_redir:
+            match = re.search(r'__tfw=([^;\s]+)', uri)
+            self.assertTrue(match, "Redirection mark is missing in Location header")
```

Comment:
So we got a rmark in the JSCH, but never used that, so atempt accounting doesn't work in JSCH? Is this how the browsers behave? If 'no' - then the code must be reworked to mimic behaviour of the browsers. If `yes` - then we probably should set the correct path (including `rmark`) in the JS. Anyway, a comment is required here.